src/Sparse/IndexCalculus.jl

```diff
@@ -0,0 +1,358 @@
+"""
+function IdxCalc(a::T, b::T, F=parent(a)) where T<:Union{gfp_elem, gfp_fmpz_elem} #RingElem better?
+    @assert parent(a) === parent(b)
+    b==1 && return fmpz(0)
+    b==a && return fmpz(1)
+    set_attribute!(F, :a=>a)
+    #typeof(get_attribute(F, :Logdict))==Nothing || @goto Logdict
+    #typeof(get_attribute(F, :RelMat))==Nothing || @goto RelMat
+    if typeof(get_attribute(F, :RelMat))==Nothing
+        sieve(F)
+    end
+    if typeof(get_attribute(F, :Logdict))==Nothing
+        p = get_attribute(F, :p)
+        _modulus = div((p-1),2)
+        two = fmpz(2)
+        #F2 = GF(_modulus)
+        F2 = ResidueRing(ZZ, _modulus) #change it when prepro fixed
+        set_attribute!(F, :F2=>F2)
+        c, u, v = gcdx(two, _modulus)
+        c == 1 || (@error "FB_LOGS: 2 ,(p-1)/2 not coprime")
+        set_attribute!(F, :u=>u, :v=>v)
+        set_attribute!(F, :_modulus=>_modulus)
+        #Preprocessing
+        A = change_base_ring(F2, get_attribute(F,:RelMat))
+        TA = transpose(A)
+        A, TA = sp_prepro(A, TA, get_attribute(F, :fb_length))
+        #Wiedemann + dict with logs of FB
+        log_dict(F, A, TA)
+    end
+    logb = log(F, b)
+    if logb == 0
+        return logb, F
+    end
+    if a != get_attribute(F, :p_elem)
+        p = get_attribute(F, :p)
+        loga = log(F, a)
+        logb = solvemod(loga, logb, p-1)
+    end
+    return logb, F 
+end
+
+function disc_log(a, b, F=parent(a)) #p prime
+    @assert parent(a) === parent(b)
+    b==1 && return fmpz(0)
+    b==a && return fmpz(1)
+    p = characteristic(F)
+    if log(p)/log(10)<13
+        @info "only bsgs used"
+        return disc_log_bs_gs(a,b,p-1), F #or ph -> compare time
+    end
+    d,t = Hecke.factor_trial_range(p-1)
+    d = sort(d)
+    small_prod = t
+    rest = p-1
+    for (k,v) in d
+        pow_ = k^v
+        divexact!(rest, pow_)
+        mul!(small_prod, small_prod, pow_)
+    end
+    h = maximum(keys(d))
+    if rest == 1
+        rest = h^d[h]
+        divexact!(small_prod, rest)
+    end
+    set_attribute!(F, :small_prod=>small_prod)
+    set_attribute!(F, :rest=>rest)
+    if small_prod == 2
+        return IdxCalc(a, b)
+    end
+    @assert log(small_prod)/log(10) < 13
+    a1 = a^rest #primitive element for small_prod
+    a2 = a^small_prod #primitive element for rest
+    b1 = b^rest
+    b2 = b^small_prod
+    g1 = disc_log_bs_gs(a1,b1,small_prod)
+    if log(rest)/log(10)<13
+        g2 = disc_log_bs_gs(a2,b2,rest)
+        @info "bsgs for rest"
+        return crt(g1, small_prod, g2, rest), F
+    else
+        @info "IdxCalc for rest"
+        @assert log(rest)/log(10)< 21
+        @error("missing")
+        return false, F
+        #in progress
+        #return disc_log_p(a2, b2, F)
+        #return a2, b2, rest, g1
+    end 
+end
+
+
+p = rand_dec_prime(17)
+F = GF(p)
+a = rand(F)
+while a == zero(F)
+    a = rand(F)
+end
+g = rand(1:p-1)
+b = a^g
+
+log_, K = disc_log(a, b)
+a^log_ == b
```

Comment:
Good idea to add a test. But this should not be under `src/` but under `testing`. There is some explanation here: http://www.thofma.com/Hecke.jl/dev/dev/test/.